### PR TITLE
Applications: download helm dependencies with credentials

### LINKED ISSUE #10845: Application:  download helm dependencies with credentials
### Description of the feature you would like to add / User story

if the template method is `Helm` and Source is GIT. In order to install the chat we must download the chart dependencies defined in `Chart.yaml`. These dependencies may point to a repository that requires authentication.

We have to update the applicationDefinitionCRD to handle credentials for this case.

Side-Note: we do not have this issue for helm registries, as they package a chart with all its dependencies

### Solution details
 
Add a new optional field `applicationDefinition.spec.version[].TemplateCredentials`  that holds the credentials that must be used when installing the application. 

Currently only Helm as method may require credentials at install time. To keep the crd easily extendable.

`TemplateCredentials` will be a struct that holds pointer(s) to the credentials. The choice of credentials will be done according to the templating method.

### PR DESCRIPTION
**Pull request by vgramer**
**What this PR does / why we need it**:

Add a new optional field `applicationDefinition.spec.version[].DependencyCredentials` that holds the credentials that must be used when installing the application.

To keep the crd easily extendable `DependencyCredentials`  is a struct that holds pointer(s) to the credentials. The choice of credentials is made according to the templating method.

**Which issue(s) this PR fixes**:

Fixes #10845 

**What type of PR is this?**

/kind feature

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:

```release-note
Applications:  update `applicationDefinition` CRD to handle credentials at "templating" time.  It allows downloading helm dependencies from an authenticated registry when application's source is git.
```

**Documentation**:

```documentation
https://github.com/kubermatic/docs/pull/1323
```

### CODE REVIEW

File: pkg/apis/apps.kubermatic/v1/application_definition.go [hunk 1]
```diff
@@ -159,6 +159,16 @@ type ApplicationTemplate struct {
...
 	// Exactly one type of source must be defined.
 	Source ApplicationSource `json:"source"`
+
+	// TemplateCredentials hodls the credentials that may need for templating application.
+	TemplateCredentials *TemplateCredentials `json:"templateCredentials,omitempty"`
+}
+
+type TemplateCredentials struct {
+	// HelmCredentials holds the ref to the secret with helm credentials needed to build helm dependencies.
+	// It's optional and only makes sense make application source is not of type Helm (e.g. git). If  application source is Helm then dependencies are already packaged in chart archive.
```
Comment (rastislavs) [inline pkg/apis/apps.kubermatic/v1/application_definition.go]:
```suggestion
	// It's optional and only makes sense if application source is not of type Helm (e.g. git). If application source is Helm then dependencies are already packaged in chart archive.
```

File: pkg/apis/apps.kubermatic/v1/application_definition.go [hunk 2]
```diff
@@ -159,6 +159,16 @@ type ApplicationTemplate struct {
 	// Defined how the source of the application (e.g Helm chart) is retrieved.
 	// Exactly one type of source must be defined.
 	Source ApplicationSource `json:"source"`
+
+	// TemplateCredentials hodls the credentials that may need for templating application.
```
Comment (SimonTheLeg) [inline pkg/apis/apps.kubermatic/v1/application_definition.go]:
```suggestion
	// TemplateCredentials holds the credentials that may be needed for templating the application.
```

File: pkg/apis/apps.kubermatic/v1/application_definition.go [hunk 3]
```diff
@@ -159,6 +159,16 @@ type ApplicationTemplate struct {
...
+	TemplateCredentials *TemplateCredentials `json:"templateCredentials,omitempty"`
+}
+
+type TemplateCredentials struct {
+	// HelmCredentials holds the ref to the secret with helm credentials needed to build helm dependencies.
+	// It's optional and only makes sense if application source is not of type Helm (e.g. git). If  application source is Helm then dependencies are already packaged in chart archive.
+	// It's Either username / Password or registryConfigFile can be defined.
```
Comment (SimonTheLeg) [inline pkg/apis/apps.kubermatic/v1/application_definition.go]:
```suggestion
	// It's either username / password or a registryConfigFile can be defined.
```

File: pkg/apis/apps.kubermatic/v1/application_definition.go [hunk 4]
```diff
@@ -159,6 +159,16 @@ type ApplicationTemplate struct {
...
 	// Exactly one type of source must be defined.
 	Source ApplicationSource `json:"source"`
+
+	// TemplateCredentials hodls the credentials that may need for templating application.
+	TemplateCredentials *TemplateCredentials `json:"templateCredentials,omitempty"`
+}
+
+type TemplateCredentials struct {
+	// HelmCredentials holds the ref to the secret with helm credentials needed to build helm dependencies.
+	// It's optional and only makes sense if application source is not of type Helm (e.g. git). If  application source is Helm then dependencies are already packaged in chart archive.
```
Comment (SimonTheLeg) [inline pkg/apis/apps.kubermatic/v1/application_definition.go]: After thinking about this, I think I would prefer that the name of the field includes the word `dependency`. My though behind this is to stress the function of the object. Additionaly I cannot foresee another use-case for this field so I think this name should be stable.
How would you feel about the name DependencyCredentials?

Comment (SimonTheLeg) [inline pkg/apis/apps.kubermatic/v1/application_definition.go]:
```suggestion
	// It is not required when using helm as a source, as dependencies are already prepackaged in this case
```

File: pkg/applications/providers/source/helm.go [hunk 5]
```diff
@@ -54,7 +54,7 @@ func (h HelmSource) DownloadSource(destination string) (string, error) {
 	}
 	defer util.CleanUpHelmTempDir(helmCacheDir, h.Log)
 
-	auth, err := util.HelmAuthFromCredentials(h.Ctx, h.SeedClient, path.Join(helmCacheDir, "reg-creg"), h.SecretNamespace, h.Source)
+	auth, err := util.HelmAuthFromCredentials(h.Ctx, h.SeedClient, path.Join(helmCacheDir, "reg-creg"), h.SecretNamespace, h.Source.Credentials)
```
Comment (SimonTheLeg) [inline pkg/applications/providers/source/helm.go:57]: I have question regarding the two calls to util.HelmAuthFromCredentials. The other one being this here https://github.com/vgramer/kubermatic/blob/67826b8392a8ecde3e9898357eae5e3c31176d0f/pkg/applications/providers/template/helm.go#L68

I see that in one of the invocations we check if the credentials is set, while in the other we don't. Is there anything to it? 

I am thinking that in the other one we avoid unnecessary calls by checking first. Would we not want a nil-check on h.Source.Credentials here as well?

**Review by SimonTheLeg — COMMENTED** [review]
One overall comment: I think we will also need a PR in dashboard to add API capabilities to it. Otherwise these fields will not be exposed.

Alternatively create a new ticket to add this functionality into the api. I'm fine with both, just want to mention so we don't forget about it.

File: pkg/applications/providers/source/helm.go (hunk 5 above)
Comment (vgramer) [reply pkg/applications/providers/source/helm.go:57]: the `util.HelmAuthFromCredentials` function takes a pointer to `*appskubermaticv1.HelmCredentials` and handle nil case. 
As` h.Source.Credentials`'s type is `*appskubermaticv1.HelmCredentials` we don't need to check if it is nil.

Comment (SimonTheLeg) [reply pkg/applications/providers/source/helm.go:57]: Okay gotcha and in the other one we need the check so we can be sure that TemplateCredentials.HelmCredentials does not panic. Makes sense.

**Comment by vgramer** [comment]
> One overall comment: I think we will also need a PR in dashboard to add API capabilities to it. Otherwise these fields will not be exposed.
> 
> Alternatively create a new ticket to add this functionality into the api. I'm fine with both, just want to mention so we don't forget about it.

I have created the issue kubermatic/dashboard#5385 to update api.

**Comment by SimonTheLeg** [comment]
> I have created the issue [kubermatic/dashboard#5385](https://github.com/kubermatic/dashboard/issues/5385) to update api.

gotcha. I have added it to the epic

**Review by SimonTheLeg — APPROVED** [review]

**Pull request merged**